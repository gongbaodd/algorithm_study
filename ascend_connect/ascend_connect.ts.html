<!DOCTYPE html>
<html><head><title>ascend_connect/ascend_connect.ts
 - Stryker report</title>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link rel="stylesheet" href="../strykerResources/bootstrap/css/bootstrap.min.css">
<link rel="stylesheet" href="../strykerResources/highlightjs/styles/default.css">
<link rel="stylesheet" href="../strykerResources/stryker/stryker.css"></head>
<body><img class="stryker-image" alt="Stryker" src="../strykerResources/stryker/stryker-80x80.png">
<div class="container-fluid"><div class="row"><div class="col-lg-12"><h1 class="display-4">ascend_connect/ascend_connect.ts
 
<small class="text-muted">- Stryker report</small></h1></div></div>
<ol class="breadcrumb"><li class="breadcrumb-item"><a href="../index.html">All files</a></li>
<li class="breadcrumb-item active">ascend_connect/ascend_connect.ts</li></ol>
<div class="row"><div class="col-lg-12"><div class="row"><div class="col-sm-11"><table class="table table-sm table-hover table-bordered table-no-top"><thead><tr><th style="width: 20%"><div><span>File / Directory</span></div></th>
<th colspan="2"><div><span>Mutation score</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Killed</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Survived</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Timeout</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># No coverage</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Runtime errors</span></div></th>
<th class="rotate text-center" style="width: 50px"><div><span># Transpile errors</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total detected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total undetected</span></div></th>
<th class="rotate rotate-width-70 text-center" style="width: 70px"><div><span>Total mutants</span></div></th></tr></thead>
<tbody><tr><td>ascend_connect/ascend_connect.ts</td>
<td><div class="progress"><div class="progress-bar bg-warning" role="progressbar" aria-valuenow="77.27" aria-valuemin="0" aria-valuemax="100" style="width: 77.27%;">77.27
%</div></div></td>
<th class="text-center text-warning" style="width: 10%">77.27</th>
<td class="text-center">9</td>
<td class="text-center">5</td>
<td class="text-center">8</td>
<td class="text-center">0</td>
<td class="text-center">16</td>
<td class="text-center">0</td>
<th class="text-center">17</th>
<th class="text-center">5</th>
<th class="text-center">38</th></tr>
</tbody></table></div></div>
<div class="row"><div class="col-lg-7"><div class="row legend"><form class="col-md-12" novalidate="novalidate">
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" checked value="2" type="checkbox">
Survived
 
(5)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="1" type="checkbox">
Killed
 
(9)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="3" type="checkbox">
TimedOut
 
(8)</label></div>
<div class="form-check form-check-inline"><label class="form-check-label"><input class="form-check-input stryker-display" value="4" type="checkbox">
RuntimeError
 
(16)</label></div>

<a href="#" class="stryker-collapse-expand-all">Expand all</a></form></div>
<pre><code class="lang-javascript"><span class="bg-null">import { LNode } from &quot;../lib&quot;;

export function ascendConnect(head1: LNode, head2: LNode): LNode </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="0"><span class="badge badge-secondary">0</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="0">{}</span><span class="stryker-original-code" data-mutant="0"><span class="bg-null">{
  if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="1"><span class="badge badge-secondary">1</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="1">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="2"><span class="badge badge-danger">2</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="2">false</span><span class="stryker-original-code" data-mutant="1"><span class="stryker-original-code" data-mutant="2"><span class="bg-danger-light">head1.next </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="3"><span class="badge badge-success">3</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="3">!==</span><span class="stryker-original-code" data-mutant="3"><span class="bg-danger-light">===</span></span><span class="bg-danger-light"> null</span></span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="4"><span class="badge badge-danger">4</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="4">{}</span><span class="stryker-original-code" data-mutant="4"><span class="bg-danger-light">{
    return head2;
  }</span></span><span class="bg-null">

  if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="5"><span class="badge badge-secondary">5</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="5">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="6"><span class="badge badge-danger">6</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="6">false</span><span class="stryker-original-code" data-mutant="5"><span class="stryker-original-code" data-mutant="6"><span class="bg-danger-light">head2.next </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="7"><span class="badge badge-secondary">7</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="7">!==</span><span class="stryker-original-code" data-mutant="7"><span class="bg-danger-light">===</span></span><span class="bg-danger-light"> null</span></span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="8"><span class="badge badge-danger">8</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="8">{}</span><span class="stryker-original-code" data-mutant="8"><span class="bg-danger-light">{
    return head1;
  }</span></span><span class="bg-null">

  let pre1 = head1;
  let cur1 = pre1.next;
  let cur2: LNode | null = head2.next;

  while (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="WhileStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="9"><span class="badge badge-secondary">9</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="9">false</span><span class="stryker-original-code" data-mutant="9"><span class="bg-null">cur2 </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="10"><span class="badge badge-secondary">10</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="10">||</span><span class="stryker-original-code" data-mutant="10"><span class="bg-null">&amp;&amp;</span></span><span class="bg-null"> cur1</span></span><span class="bg-null">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="11"><span class="badge badge-warning">11</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="11">{}</span><span class="stryker-original-code" data-mutant="11"><span class="bg-warning-light">{
    if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="12"><span class="badge badge-success">12</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="12">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="13"><span class="badge badge-secondary">13</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="13">false</span><span class="stryker-original-code" data-mutant="12"><span class="stryker-original-code" data-mutant="13"><span class="bg-warning-light">(</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="14"><span class="badge badge-secondary">14</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="14">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="15"><span class="badge badge-secondary">15</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="15">false</span><span class="stryker-original-code" data-mutant="14"><span class="stryker-original-code" data-mutant="15"><span class="bg-warning-light">cur1.data </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="16"><span class="badge badge-secondary">16</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="16">&amp;&amp;</span><span class="stryker-original-code" data-mutant="16"><span class="bg-warning-light">||</span></span><span class="bg-warning-light"> 0</span></span></span><span class="bg-warning-light">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: Survived" data-mutant-status-annotation="danger" data-mutant-status="2" data-mutant="17"><span class="badge badge-danger">17</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="17">&lt;=</span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="18"><span class="badge badge-warning">18</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="18">&gt;=</span><span class="stryker-original-code" data-mutant="17"><span class="stryker-original-code" data-mutant="18"><span class="bg-danger-light">&lt;</span></span></span><span class="bg-warning-light"> (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="19"><span class="badge badge-secondary">19</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="19">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="20"><span class="badge badge-secondary">20</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="20">true</span><span class="stryker-original-code" data-mutant="19"><span class="stryker-original-code" data-mutant="20"><span class="bg-warning-light">cur2.data </span><a href="#" class="stryker-mutant-button" tabindex="0" title="BinaryExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="21"><span class="badge badge-secondary">21</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="21">&amp;&amp;</span><span class="stryker-original-code" data-mutant="21"><span class="bg-warning-light">||</span></span><span class="bg-warning-light"> 0</span></span></span><span class="bg-warning-light">)</span></span></span><span class="bg-warning-light">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="22"><span class="badge badge-warning">22</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="22">{}</span><span class="stryker-original-code" data-mutant="22"><span class="bg-warning-light">{
      const next2: LNode | null = </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="23"><span class="badge badge-secondary">23</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="23">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="24"><span class="badge badge-secondary">24</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="24">true</span><span class="stryker-original-code" data-mutant="23"><span class="stryker-original-code" data-mutant="24"><span class="bg-warning-light">cur2?.next ?? null</span></span></span><span class="bg-warning-light">;

      </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="25"><span class="badge badge-success">25</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="25">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="26"><span class="badge badge-success">26</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="26">true</span><span class="stryker-original-code" data-mutant="25"><span class="stryker-original-code" data-mutant="26"><span class="bg-warning-light">pre1.next = cur2</span></span></span><span class="bg-warning-light">;
      </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="27"><span class="badge badge-success">27</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="27">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="28"><span class="badge badge-success">28</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="28">true</span><span class="stryker-original-code" data-mutant="27"><span class="stryker-original-code" data-mutant="28"><span class="bg-warning-light">cur2.next = cur1</span></span></span><span class="bg-warning-light">;
      </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="29"><span class="badge badge-warning">29</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="29">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="30"><span class="badge badge-warning">30</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="30">true</span><span class="stryker-original-code" data-mutant="29"><span class="stryker-original-code" data-mutant="30"><span class="bg-warning-light">cur2 = next2</span></span></span><span class="bg-warning-light">;
    }</span></span><span class="bg-warning-light">

    if (</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="31"><span class="badge badge-success">31</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="31">true</span><a href="#" class="stryker-mutant-button" tabindex="0" title="IfStatement" data-content="status: RuntimeError" data-mutant-status-annotation="secondary" data-mutant-status="4" data-mutant="32"><span class="badge badge-secondary">32</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="32">false</span><span class="stryker-original-code" data-mutant="31"><span class="stryker-original-code" data-mutant="32"><span class="bg-warning-light">cur1.next</span></span></span><span class="bg-warning-light">) </span><a href="#" class="stryker-mutant-button" tabindex="0" title="Block" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="33"><span class="badge badge-warning">33</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="33">{}</span><span class="stryker-original-code" data-mutant="33"><span class="bg-warning-light">{
      </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="34"><span class="badge badge-success">34</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="34">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: Killed" data-mutant-status-annotation="success" data-mutant-status="1" data-mutant="35"><span class="badge badge-success">35</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="35">true</span><span class="stryker-original-code" data-mutant="34"><span class="stryker-original-code" data-mutant="35"><span class="bg-warning-light">pre1 = cur1</span></span></span><span class="bg-warning-light">;
      </span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="36"><span class="badge badge-warning">36</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="36">false</span><a href="#" class="stryker-mutant-button" tabindex="0" title="ConditionalExpression" data-content="status: TimedOut" data-mutant-status-annotation="warning" data-mutant-status="3" data-mutant="37"><span class="badge badge-warning">37</span></a><span class="badge badge-info stryker-mutant-replacement" hidden="hidden" data-mutant="37">true</span><span class="stryker-original-code" data-mutant="36"><span class="stryker-original-code" data-mutant="37"><span class="bg-warning-light">cur1 = cur1.next</span></span></span><span class="bg-warning-light">;
    }</span></span><span class="bg-warning-light">
  }</span></span><span class="bg-null">

  return head1;
}</span></span><span class="bg-null">
</code></pre></div>
<div class="col-lg-5"><table class="table table-sm table-hover mutant-table"><thead><tr><th>#</th>
<th>Mutator</th>
<th>State</th>
<th>Location</th>
<th>Original</th>
<th>Replacement</th></tr></thead>
<tbody><tr><th>0</th>
<td>Block</td>
<td>RuntimeError</td>
<td>2
:
65</td>
<td><code>{
   ...  ;
}</code></td>
<td><code>{}</code></td></tr>
<tr><th>1</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>3
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>2</th>
<td>IfStatement</td>
<td>Survived</td>
<td>3
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>3</th>
<td>BinaryExpression</td>
<td>Killed</td>
<td>3
:
17</td>
<td><code>===</code></td>
<td><code>!==</code></td></tr>
<tr><th>4</th>
<td>Block</td>
<td>Survived</td>
<td>3
:
27</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>5</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>7
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>6</th>
<td>IfStatement</td>
<td>Survived</td>
<td>7
:
6</td>
<td><code>.     ===</code></td>
<td><code></code></td></tr>
<tr><th>7</th>
<td>BinaryExpression</td>
<td>RuntimeError</td>
<td>7
:
17</td>
<td><code>===</code></td>
<td><code>!==</code></td></tr>
<tr><th>8</th>
<td>Block</td>
<td>Survived</td>
<td>7
:
27</td>
<td><code>{
   ...;
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>9</th>
<td>WhileStatement</td>
<td>RuntimeError</td>
<td>15
:
9</td>
<td><code>&&</code></td>
<td><code></code></td></tr>
<tr><th>10</th>
<td>BinaryExpression</td>
<td>RuntimeError</td>
<td>15
:
14</td>
<td><code>&&</code></td>
<td><code>||</code></td></tr>
<tr><th>11</th>
<td>Block</td>
<td>TimedOut</td>
<td>15
:
23</td>
<td><code>{
   ...}
  }</code></td>
<td><code>{}</code></td></tr>
<tr><th>12</th>
<td>IfStatement</td>
<td>Killed</td>
<td>16
:
8</td>
<td><code>(    ...||  )</code></td>
<td><code></code></td></tr>
<tr><th>13</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>16
:
8</td>
<td><code>(    ...||  )</code></td>
<td><code></code></td></tr>
<tr><th>14</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>16
:
9</td>
<td><code>.     ||</code></td>
<td><code></code></td></tr>
<tr><th>15</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>16
:
9</td>
<td><code>.     ||</code></td>
<td><code></code></td></tr>
<tr><th>16</th>
<td>BinaryExpression</td>
<td>RuntimeError</td>
<td>16
:
19</td>
<td><code>||</code></td>
<td><code>&&</code></td></tr>
<tr><th>17</th>
<td>BinaryExpression</td>
<td>Survived</td>
<td>16
:
25</td>
<td><code><</code></td>
<td><code><=</code></td></tr>
<tr><th>18</th>
<td>BinaryExpression</td>
<td>TimedOut</td>
<td>16
:
25</td>
<td><code><</code></td>
<td><code>>=</code></td></tr>
<tr><th>19</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>16
:
28</td>
<td><code>.     ||</code></td>
<td><code></code></td></tr>
<tr><th>20</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>16
:
28</td>
<td><code>.     ||</code></td>
<td><code></code></td></tr>
<tr><th>21</th>
<td>BinaryExpression</td>
<td>RuntimeError</td>
<td>16
:
38</td>
<td><code>||</code></td>
<td><code>&&</code></td></tr>
<tr><th>22</th>
<td>Block</td>
<td>TimedOut</td>
<td>16
:
45</td>
<td><code>{
   ...    }</code></td>
<td><code>{}</code></td></tr>
<tr><th>23</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>17
:
34</td>
<td><code>?.     ??</code></td>
<td><code></code></td></tr>
<tr><th>24</th>
<td>ConditionalExpression</td>
<td>RuntimeError</td>
<td>17
:
34</td>
<td><code>?.     ??</code></td>
<td><code></code></td></tr>
<tr><th>25</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>19
:
6</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>26</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>19
:
6</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>27</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>20
:
6</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>28</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>20
:
6</td>
<td><code>.     =</code></td>
<td><code></code></td></tr>
<tr><th>29</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>21
:
6</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>30</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>21
:
6</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>31</th>
<td>IfStatement</td>
<td>Killed</td>
<td>24
:
8</td>
<td><code>.</code></td>
<td><code></code></td></tr>
<tr><th>32</th>
<td>IfStatement</td>
<td>RuntimeError</td>
<td>24
:
8</td>
<td><code>.</code></td>
<td><code></code></td></tr>
<tr><th>33</th>
<td>Block</td>
<td>TimedOut</td>
<td>24
:
19</td>
<td><code>{
   ...    }</code></td>
<td><code>{}</code></td></tr>
<tr><th>34</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>25
:
6</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>35</th>
<td>ConditionalExpression</td>
<td>Killed</td>
<td>25
:
6</td>
<td><code>=</code></td>
<td><code></code></td></tr>
<tr><th>36</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>26
:
6</td>
<td><code>=     .</code></td>
<td><code></code></td></tr>
<tr><th>37</th>
<td>ConditionalExpression</td>
<td>TimedOut</td>
<td>26
:
6</td>
<td><code>=     .</code></td>
<td><code></code></td></tr></tbody></table></div></div></div></div>
<footer>Generated with stryker-html-reporter generator. Visit the 
<a href="https://stryker-mutator.io" target="_blank">Stryker website</a></footer></div>
<script src="../strykerResources/popper.js/dist/umd/popper.min.js" defer="defer"></script>
<script src="../strykerResources/jquery/dist/jquery.slim.min.js" defer="defer"></script>
<script src="../strykerResources/bootstrap/js/bootstrap.min.js" defer="defer"></script>
<script src="../strykerResources/stryker.js" defer="defer"></script></body></html>