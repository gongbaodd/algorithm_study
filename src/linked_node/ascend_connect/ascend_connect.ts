import { LNode } from "../lib";

export function ascendConnect(head1: LNode, head2: LNode): LNode {
  if (head1.next === null) {
    return head2;
  }

  if (head2.next === null) {
    return head1;
  }

  let pre1 = head1;
  let cur1 = pre1.next;
  let cur2: LNode | null = head2.next;

  while (cur2 && cur1) {
    if (cur1.data < cur2.data) {
      const next2: LNode | null = cur2?.next ?? null;

      pre1.next = cur2;
      cur2.next = cur1;
      cur2 = next2;
    }

    if (cur1.next) {
      pre1 = cur1;
      cur1 = cur1.next;
    }
  }

  return head1;
}
