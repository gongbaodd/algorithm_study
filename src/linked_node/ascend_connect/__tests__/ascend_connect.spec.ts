import { ascendConnect } from "../ascend_connect";
import { init } from "../../lib";

describe("ascend connect", () => {
  describe("chain1 is longer than chain2", () => {
    describe("chain1 = [5, 3, 1], chain2 = [4, 2]", () => {
      it("should return [5, 4, 3, 2, 1]", () => {
        expect(ascendConnect(init([5, 3, 1]), init([4, 2]))).toMatchSnapshot();
      });
    });

    describe("chain1 = [5, 3, 1], chain2 = [6, 2]", () => {
      it("should return [6, 5, 3, 2, 1]", () => {
        expect(ascendConnect(init([5, 3, 1]), init([6, 2]))).toMatchSnapshot();
      });
    });
  });

  describe("chain1 is shorter than chain2", () => {
    describe("chain1 = [5, 1], chain2 = [6, 4, 2]", () => {
      it("should return [6, 5, 4, 2, 1]", () => {
        expect(ascendConnect(init([5, 1]), init([6, 4, 2]))).toMatchSnapshot();
      });
    });

    describe("chain1 = [3, 1], chain2 = [6, 4, 2]", () => {
      it("should return [6, 4, 3, 2, 1]", () => {
        expect(ascendConnect(init([3, 1]), init([6, 4, 2]))).toMatchSnapshot();
      });
    });
  });
});
