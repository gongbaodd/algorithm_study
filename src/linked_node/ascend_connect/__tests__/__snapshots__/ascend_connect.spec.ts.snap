// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ascend connect chain1 is longer than chain2 chain1 = [5, 3, 1], chain2 = [4, 2] should return [5, 4, 3, 2, 1] 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 5,
    "next": LNode {
      "data": 4,
      "next": LNode {
        "data": 3,
        "next": LNode {
          "data": 2,
          "next": LNode {
            "data": 1,
            "next": null,
          },
        },
      },
    },
  },
}
`;

exports[`ascend connect chain1 is longer than chain2 chain1 = [5, 3, 1], chain2 = [6, 2] should return [6, 5, 3, 2, 1] 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 6,
    "next": LNode {
      "data": 5,
      "next": LNode {
        "data": 3,
        "next": LNode {
          "data": 2,
          "next": LNode {
            "data": 1,
            "next": null,
          },
        },
      },
    },
  },
}
`;

exports[`ascend connect chain1 is shorter than chain2 chain1 = [3, 1], chain2 = [6, 4, 2] should return [6, 4, 3, 2, 1] 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 6,
    "next": LNode {
      "data": 3,
      "next": LNode {
        "data": 2,
        "next": LNode {
          "data": 1,
          "next": null,
        },
      },
    },
  },
}
`;

exports[`ascend connect chain1 is shorter than chain2 chain1 = [5, 1], chain2 = [6, 4, 2] should return [6, 5, 4, 2, 1] 1`] = `
LNode {
  "data": 0,
  "next": LNode {
    "data": 6,
    "next": LNode {
      "data": 5,
      "next": LNode {
        "data": 2,
        "next": LNode {
          "data": 1,
          "next": null,
        },
      },
    },
  },
}
`;
